onnxruntime/core/graph/graph_utils.cc

```diff
@@ -349,25 +309,111 @@ bool IsOutputUsed(const Node& node, int index) {
   return false;
 }
 
-bool RemoveNode(Graph& graph, Node& node) {
-  // Cannot remove a node with implicit inputs, whose output is also a graph output,
-  // or with more than one of its outputs as input to downstream Operators.
-  if (!node.ImplicitInputDefs().empty() ||
-      graph.IsNodeOutputsInGraphOutputs(node) || !IsOnlyOneOutputUsed(node)) {
+bool CanRemoveNodeAndMergeEdges(const Graph& graph, const Node& node) {
+  const std::string* output_name = nullptr;
+  if (!IsOnlyOneOutputUsed(graph, node, output_name)) {
+    return false;
+  }
+
+  // TODO: Currently we remove the node and use the input name from the node being removed.
+  // It may also be possible to instead update an upstream node to use the output name from the node being removed.
+  // This would allow removal of a node that is providing a graph output, as that output name would come from updating
+  // the upstream node. This should also enable removal if CanUpdateImplicitInputNameInSubgraphs returns false.
+
+  if (!graph.GetNodeOutputsInGraphOutputs(node).empty()) {
     return false;
   }
 
+  bool can_remove = false;
+  const std::string* new_name = nullptr;
+
   if (node.GetInputEdgesCount() == 1) {
-    // If there is a single input edge from another node (initializers are not connected with edges to nodes).
+    // we will merge the single input edge with the edges for the output that is used
```

Comment:
Similar to the comment in the .h, I think we should mention that if there is a single input edge, we use that one (but that does not mean that the node has no other inputs).

Reply:
Updated comment